### PR TITLE
Decouple utilities from `LightningLoggerBase`

### LINKED ISSUE #11237: Move protected static methods in `LightningLoggerBase` to utilities file
## Proposed refactor

Move this code out of the base logger class to a utilities file: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/dedfde6859a15270385e6e751567b29ab6488e87/pytorch_lightning/loggers/base.py#L181-L280
https://github.com/PyTorchLightning/pytorch-lightning/blob/dedfde6859a15270385e6e751567b29ab6488e87/pytorch_lightning/loggers/base.py#L363-L367

### Motivation

Streamline the `LightningLoggerBase` as the core interface required for implementations. 

These methods do not need to be part of the base class, since they're not required to be called or overridden. They can be moved out of the base class into a utils file in order to still be used in various logger implementations.

Moving these methods out makes the base class easier to read & follow.

There's no deprecation required since these methods are marked private already. 

### Pitch

Move these functions out, to either `pl/loggers/utilities.py` or `pl/utilities/logger.py` 

If agreed on, this could be a Good First Issue candidate

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @awaelchli @akihironitta @rohitgr7 @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy

### PR DESCRIPTION
**Pull request by akashkw**
## What does this PR do?

Fixes #11237

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by akihironitta** [comment]
Hi @akashkw, thank you for working on this! Let me know if you need help :]

**Review by tchaton — APPROVED** [review]
LGTM ! Thanks for contributing!

File: pytorch_lightning/loggers/wandb.py [hunk 1]
```diff
@@ -356,16 +357,16 @@ def watch(self, model: nn.Module, log: str = "gradients", log_freq: int = 100, l
 
     @rank_zero_only
     def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
-        params = self._convert_params(params)
-        params = self._flatten_dict(params)
-        params = self._sanitize_callable_params(params)
+        params = _convert_params(params)
+        params = _flatten_dict(params)
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/wandb.py:361]: off-topic. But I think wandb supports logging nested dict for hparams. Maybe we shouldn't force the behavior here.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -195,8 +197,8 @@ def log_hyperparams(
             self.hparams.update(params)
 
         # format params into the suitable for tensorboard
-        params = self._flatten_dict(params)
-        params = self._sanitize_params(params)
```
Comment (justusschock) [inline pytorch_lightning/loggers/tensorboard.py:199]: Why not using self here anymore?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loggers/tensorboard.py [hunk 3]
```diff
@@ -195,8 +197,8 @@ def log_hyperparams(
...
+        params = _flatten_dict(params)
+        params = TensorBoardLogger._sanitize_params(params)
```
Comment (ananthsub) [inline pytorch_lightning/loggers/tensorboard.py:201]:
```suggestion
        params = self._sanitize_params(params)
```

File: pytorch_lightning/utilities/logger.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+"""Utilities for loggers."""
+
+from argparse import Namespace
+from typing import Any, Dict, Generator, List, MutableMapping, Optional, Union
+
+import numpy as np
+import torch
+
+
+def _convert_params(params: Union[Dict[str, Any], Namespace]) -> Dict[str, Any]:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/logger.py:23]: please add unit tests for these functions outside of the integration tests with the Loggers

File: pytorch_lightning/utilities/logger.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        elif type(params[k]) not in [bool, int, float, str, torch.Tensor]:
+            params[k] = str(params[k])
+    return params
+
+
+def _add_prefix(metrics: Dict[str, float], prefix: str, logger_join_char: str) -> Dict[str, float]:
+    if prefix:
+        metrics = {f"{prefix}{logger_join_char}{k}": v for k, v in metrics.items()}
+
+    return metrics
```
Comment (ananthsub) [inline pytorch_lightning/utilities/logger.py:130]:
```suggestion
def _add_prefix(metrics: Dict[str, float], prefix: str, separator: str) -> Dict[str, float]:
    if prefix:
        metrics = {f"{prefix}{separator}{k}": v for k, v in metrics.items()}

    return metrics
```

**Review by ananthsub — COMMENTED** [review]
@awaelchli do you know why some loggers have LOGGER_JOIN_CHAR while others implement the group separator property? are these intended to be used differently?  https://github.com/PyTorchLightning/pytorch-lightning/blob/1a2536361749d13e3f30cec11c7da11747f425ab/pytorch_lightning/loggers/base.py#L348-L351

vs 

[LOGGER_JOIN_CHAR](https://github.com/PyTorchLightning/pytorch-lightning/blob/1a2536361749d13e3f30cec11c7da11747f425ab/pytorch_lightning/loggers/comet.py#L129)

File: tests/loggers/test_wandb.py [hunk 6]
```diff
@@ -300,9 +301,9 @@ def wrapper_something():
     params.wrapper_something_wo_name = lambda: lambda: "1"
     params.wrapper_something = wrapper_something
 
-    params = WandbLogger._convert_params(params)
-    params = WandbLogger._flatten_dict(params)
-    params = WandbLogger._sanitize_callable_params(params)
+    params = _convert_params(params)
+    params = _flatten_dict(params)
+    params = _sanitize_callable_params(params)
```
Comment (akihironitta) [inline tests/loggers/test_wandb.py:306]: Since this test isn't related to wandb at all, let's move it to `tests/utilities/test_logger.py` and add other unit tests as @\ananthsub suggests in https://github.com/PyTorchLightning/pytorch-lightning/pull/11484/files#r789031440?

File: pytorch_lightning/loggers/tensorboard.py (hunk 2 above)
Comment (akashkw) [reply pytorch_lightning/loggers/tensorboard.py:199]: self is correct here, it has been fixed now.

**Comment by awaelchli** [comment]
> @awaelchli do you know why some loggers have LOGGER_JOIN_CHAR while others implement the group separator property? are these intended to be used differently?

When we introduced group-separator, I guess nobody remembered we had a LOGGER_JOIN_CHAR. It is currently only used by the the global prefixing and has no effect for most loggers that set prefix="" in the init. It could use group_separator as well. 

- LOGGER_JOIN_CHAR is more practical as it can be changed, without overriding the logger class (is it meant to be??)
- group-separator is a method and needs to be overridden to be customized. 
- allowing it to be changed freely may not be the best, as some loggers may not support or not properly interpret these separators (e.g. tensorboard)

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/logger.py (hunk 4 above)
Comment (akashkw) [reply pytorch_lightning/utilities/logger.py:23]: Unit tests have been added! I moved the wandb unit test to this new file as well, as @akihironitta suggests here https://github.com/PyTorchLightning/pytorch-lightning/pull/11484#discussion_r789869762

File: tests/loggers/test_wandb.py (hunk 6 above)
Comment (akashkw) [reply tests/loggers/test_wandb.py]: Done!

**Review by ananthsub — APPROVED** [review]

**Review by daniellepintz — APPROVED** [review]
Looks great, nice work Akash! Btw the remaining test failure seems to be an unrelated issue affecting multiple PRs

**Review by Borda — APPROVED** [review]

**Comment by daniellepintz** [comment]
Can someone with access please rerun the TPU test and enable auto-merge on this?
